### PR TITLE
Add adr0001 - why project pravic?

### PR DESCRIPTION
**Pull request by jillr**
##### SUMMARY
Initialize ADR directory
Add an architectural decision record with the rational behind the project

##### ISSUE TYPE

- Docs Pull Request

##### COMPONENT NAME
adr

##### ADDITIONAL INFORMATION

This ADR should be circulated widely to the community before merging

### CODE REVIEW

File: docs/adr/0001-why-pravic.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+This work would additionally allow us to better enable machine-driven content creation and discovery and reduce the time to market and maintenance burden for new content.
+
+### Resource Declaration
+
+A resource definition should be the authoritative view of the state or status of a resource.  The definition provides a configuration as code view of the desired state of the resource.  This differs from Ansible's existing task-based approach to cloud automation by abstracting the individual steps necessary to achieve the desired state, which can involve many modules and tasks which must be executed in a strict order for many cloud services today.
```
Comment (gravesm) [inline docs/adr/0001-why-pravic.md:37]:
```suggestion
A resource definition should be the authoritative view of the state or status of a resource.  The definition provides a configuration as code view of the desired state of the resource.  This differs from Ansible's existing task-based approach to cloud automation by abstracting the individual steps necessary to achieve the desired state, which can involve many modules and tasks which must be executed in a strict order for many cloud services today. An example resource definition might look something like this:
```

File: docs/adr/0001-why-pravic.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+### Hosts
+
+With a web API, the state that we're managing lives on the other end of an HTTP connection. It doesn't really matter where we run our playbook. The notion of local and remote host is largely meaningless in this context and adds a maintenance burden on developers and a cognitive burden on users. This functionality provides little benefit to REST-based collections.
+
+### Task Execution
+The Ansible model where each task is executed in a separate Python process is an expensive constraint that prevents the easy reuse of connections. When a collection does all its work over HTTP, the benefits of connection reuse can be significant. Turbo Mode was designed to address this, but it is not a viable long-term solution. The maintenance burden is high. It requires extra work by plugin developers and playbook authors, and this is not something either of these personas should have to think about. It should be provided as base functionality by Ansible.
```
Comment (gravesm) [inline docs/adr/0001-why-pravic.md:27]:
```suggestion
The Ansible model where each task is executed in a separate Python process is an expensive constraint that prevents the easy reuse of connections. When a collection does all its work over HTTP, the benefits of connection reuse can be significant. [Turbo Mode](https://github.com/ansible-collections/cloud.common#ansible-turbo-module) was designed to address this, but it is not a viable long-term solution. The maintenance burden is high. It requires extra work by plugin developers and playbook authors, and this is not something either of these personas should have to think about. It should be provided as base functionality by Ansible.
```

File: docs/adr/0001-why-pravic.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+### Dependency Resolution
+
+Cloud resources commonly depend on other resources to exist or be in a specific state before they can be created.  This often causes frustration for Ansible users as they must maintain a high degree of awareness of these relationships and ensure that tasks have the appropriate checks for resource readiness and availability.
+
+Where a vendor provides a reference schema for their resources, such as in the case of the AWS Cloud Control API, we can programmatically identify these dependencies and make decisions about the order of resource actions.
+
+In the above example, the VPC must be created before any other resources and the security group must be created before the instance and the elastic cache.  We could identify this both through the resource references in the resource definition and by analyzing the AWS schemas for these resources.
+
+### Asynchronous Execution
```
Comment (gravesm) [inline docs/adr/0001-why-pravic.md:94]: I feel like this section could be a bit clearer. It switches back and forth between talking about Ansible's async execution and pravic's async execution. They're both useful, but the main selling point with pravic is that the playbook author doesn't need to think about it. Asynchronous execution isn't a playbook level feature, it's just built directly into the collection.

**Review by gravesm — APPROVED** [review]

**Comment by alinabuzachis** [comment]
recheck

**Review by alinabuzachis — APPROVED** [review]

**Comment by dmsimard** [comment]
I think I understand the use case and I want to share how I create infrastructure with ansible (sometimes), in case it could give you ideas.

TL;DR: I create myself a role or playbook and then use that as an interface to the various modules. I wouldn't really say it's declarative (I'm not interested in arguing about that) but you could say the imperative part is mostly hidden away behind an abstraction layer: the role or playbook :p

I can supply these the necessary vars (what image, ssh key, firewall rules, or virtual networks, etc) in a lot of ways, from the inventory, as extra vars or a vars file that looks like the example from this PR under ``Resource Declaration``.
Underneath, there can be conditions, assertions, handlers or async... anything that ansible makes possible.

For an example playbook that creates things on digitalocean where I also end up adding the host to the inventory such that I can deploy on top of it once it's up and available: https://github.com/ansible-community/ara-infra/blob/5bf45600148e7dfa8e1c92f801a71a0ddb9157fc/playbooks/digitalocean_infra.yaml#L78-L83

For a role example, I have an (unfortunately closed source) role called ``azure_vm`` which does something a bit like this (but with other resources, including the resource group, etc): https://github.com/ansible-collections/azure/blob/48c2c251fcc9ada0c9a39c58d34216b9fae85722/tests/integration/targets/azure_rm_virtualmachine/tasks/azure_test_minimal_manageddisk.yml#L1-L45

I have some sane defaults set up in role variables such that I don't need to specify everything and some are omitted if not provided but you get the idea.

I say this because Collections can ship roles and playbooks now, not only modules and plugins. They could be used as abstraction layers to manage infrastructure a bit like this ¯\\_(ツ)_/¯

In other words, there could be an ``aws_vm`` (or whatever, _naming is hard_) role in the collection that creates not just the VM but the things above, underneath and around it. There could be one for azure, one for gcp, one for docker or podman (that'd even take care of installing them if they're not installed?) and so on.

I don't have all the answers, it wouldn't be the fastest and it probably doesn't tick all of the boxes but could be an interesting approach, perhaps.

Of course, I like the regular ansible approach because then I have reporting on every individual playbook, host and task with ara :p

**Comment by jillr** [comment]
@dmsimard I understand where you're coming from for the use case of VMs, but public cloud is so incredibly vast we can't even keep up with modules let alone create and support roles for everything our users do on public cloud.  Generated collections have been explored by both Ansible and GCP and this helps, but it's still an ongoing game of catch-up.  Abstracting resources in this way, rather than creating a single module for every resource on every hyperscaler, gives users a more direct and faster way to use new resource types as they're made available on the hyperscaler's resource management API.

Unfortunately it's really not feasible or us to make roles for everything.  We just don't have nearly enough people (combined in the community and Ansible eng) to attempt something like this for even one hyperscaler for enough resources to be broadly useful.

**Comment by alinabuzachis** [comment]
recheck
